### ISSUE TITLE
Text layer

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#593]
## 🚀 Feature
It would be nice to have a text layer built using the [vispy text visual](http://vispy.org/visuals.html#vispy.visuals.TextVisual) so that you can display a string or list of stings overlaid on the canvas

**Comment by jni** [comment]
Tangentially related: using the vispy text visual to display keyboard shortcuts: https://github.com/napari/napari/pull/583#issuecomment-541489940

**Comment by kevinyamauchi** [comment]
@sofroniewn, this would be useful for me! I have some time this week to take a crack at it, if you haven't started already.

**Comment by sofroniewn** [comment]
Thats great - haven’t started, so go for it @kevinyamauchi.

You might want to use the Points layer as a model, as I imagine we want to support lists of text strings with a list of coordinates of the top left / center of each string. In that way it is pretty similar to the points layer

**Comment by kevinyamauchi** [comment]
@sofroniewn , I made an [initial prototype](https://github.com/kevinyamauchi/napari/commit/bcd460dfddb1ad4187175997870078a6c7a029fe) and I am a bit on the fence. Initially, I thought it would be nice to have a marker and text so it is clear what the text is referring to. In cases that we don't need the marker, we can just hide them (the visibility of the `Marker` node would be settable). Now that I see it, I'm not so sure if the markers are necessary (you can see the `add_annotatons.py` example). What do you think?

**Comment by jni** [comment]
@kevinyamauchi thanks for the link to the prototype! Note that you should feel free to submit as a WIP/draft PR (GitHub now lets you mark these "officially" as draft which prevents merging! Which is very cool), which allows us to comment directly on specific aspects of the code.

Regarding points, I am also on the fence, except to say that if you want to have points markers, you might as well subclass the Points layer?

I'm also not sure about the name, and would probably prefer to refer to it as "Text" or at best "TextAnnotations".

Finally, I think offsets are best left to automated layout tools, and instead would focus on things like "horizontalalign", "verticalalign" and "margin".

**Comment by sofroniewn** [comment]
@kevinyamauchi amazing progress! I gave `add_annotations.py` a look, and while the Points were a nice idea, taking a look at it, I think we can do without them. The points can always be added in a separate layer (and maybe linked with an automated callback if needed), and I could equally imagine someone wanting to put text near shapes instead.

Not having the points will make the UI controls much simple as you wont need properties controls for both markers and the text.

I'd like the color property of the layer to control the color of the text, and the size property to control font size. Is it possible to have different colors / text sizes for each text box? If so then that's great and we can reuse all the selection machinery inside Points to achieve that, otherwise these can be global properties.

If the markers are gone I'd also prefer it if then selection can then use the center of the text box and the height and width of the textbox - maybe computed by a (scale_factor*fontsize, scale_factor*fontsize*len(text)). I really like that you can drag the text around even now.

Even dropping the explicit points, I'd still keep a single offset vector, as it is nice to get be able to keep the coordinate data where you want it but have the text slightly differently positioned. 

Offsets on the layer model should be a vector of the same length and order as the layer shape, and if they aren't applied  you should default them to be a vector of zeros of length the layer.ndim. The reason for this is that it will handle all our nD stuff properly. You might also need to do some stuff to think about how to update the length of offsets if the dimensionality of the layer changes, but we can worry about that on the PR.

How does everyone feel about making the main data object here a tuple `(points, text)` where points is the nD array of coordinates, and text is the list of text strings. The tuple would be accessible under `layer.data`, but the individual properties could be returned from `layer.points` and `layer.text`. This would match how we handled a more complex data type in the `Shapes` layer, which is now also a tuple.

I'd also prefer the layer to be called just `Text`, which especially if we drop the Points will be more clear.

Finally no need to make the text strings editable from the GUI in this PR, but one day that might be nice.

It's a great start, and as @jni said you can make a [WIP] PR and we can continue the discussion there.

**Comment by kevinyamauchi** [comment]
Thanks for all of the feedback @sofroniewn and @jni! I didn't know that Github PRs have a "draft" mode - super cool! I have made some of the suggested changes in #600 .

I will move the main discussion points over to the PR to keep the discussion together.

**Comment by kevinyamauchi** [comment]
I am resurrecting this issue because I think we have finally restructured our `Shapes` and `Points` layers to a place where we can implement text as outlined by @jni  in https://github.com/napari/napari/pull/600#issuecomment-565241529

> (1) Points are required to have a data point, coords=, that is conceptually an npoints x ndim array.
(2) They can also have other properties, including color, size, marker, and annotations. These properties are per-point, and they can be straight up lists: add_points(coords, color=colorlist, size=sizelist, annotations=strlist).
(3) They can, however, also be columns in an dataframe (or dataframe-like object), and they can then be addressed by column name. We can call the dataframe properties. Let's say properties has column names 'annotation', 'vesicle density', 'pre/post-synaptic', 'neurotransmitter'. Then you could call add_points(coords, properties=props, size='vesicle density', marker='pre/post-synaptic', color='neurotransmitter'). The properties would then be mapped appropriately using color cycles and shape cycles, just like matplotlib/seaborn do.
(4) annotations by default would be centered on their point, but we can add an offset= argument that contains individual offsets, and the argument could also be set to "auto" to do something like ggrepel.
>
> As mentioned, the above paradigm could also be used to annotate lines, shapes, etc, and take the handling of mapping data properties to visuals away from users and into our layers, which, as the various declarative data visualization libraries have shown, is a honking good idea. All of the mappings would of course be configurable, but they would remain automatic, and the list input will always remain an option for everyone who likes doing stuff by hand.

**Proposal**
Towards this, I propose the following:

- We add text to `Points` and `Shapes`. The displayed text must be stored in `properties`.
- We create a `Text` (name can be optimized...) class that holds the relevant parameters
    - property name the text values are stored in
    - translation from the point coordinate
    - rotation around the point coordinate
    - size
    - color (I would be okay with just making it the same as the point to save complexity)
    - visibility
- We create a `TextMixIn` (name can be optimized...) that adds the text property and required events. We can then use this as a mixin with the `Shapes` and `Text` classes to add text to the layer models. We can consider doing the same for #1361 as well.

This shouldn't be too bad, as we worked out a lot of the code in #600. If this makes sense, I would start a new PR that would supersede #600. How does this sound, @napari/core-devs ?

**Comment by sofroniewn** [comment]
> We create a TextMixIn (name can be optimized...) that adds the text property and required events. We can then use this as a mixin with the Shapes and Text classes to add text to the layer models.

How about a `TextManager` class that we can put on the layer through composition - not through a mixin and inheritence (see comment in #1343 about preferring composition to inheritance). We'd then be able to access things like `layer.text.color` and `layer.text.size` instead of `layer.text_size`. Once we have our `ColorManger` class we could reuse that for `layer.text.color` too.

We could then have a `VispyTextVisual` that would render based just on that class.

One application to keep in mind are things like I put in here https://github.com/napari/napari/issues/177#issuecomment-644213821 where we have text around a bounding box for shapes.

> We can consider doing the same for #1361 as well.

I definitely want this Class to be reusable for Tracks

We should also think that it will be reusable to the "placeholder" text that @pwinston needs for #1354.

**Comment by kevinyamauchi** [comment]
@sofroniewn and I just had a conversation on slack that I have summarized below:

- We agreed that we should use a `TextManager` class that gets incorporated to the relevant layers via composition, rather than inheritance.
- methods related to the text (e.g., calculating the highlight bounding box) will be part of the `TextManager` class and the required data from the model will be passed via input arguments (e.g., the `selected_data`)

Also, an additional feature: the user will be able to pass something like an f-string as the text. For example:

```python
properties = {'point_class': ['cat', 'dog'], 'confidence': [0.95, 0.5]}
text = 'class: {point_class}, condifidence: {confidence}'
points_layer = viewer.add_points(data, properties=properties, text=text)
```
would yield text labels:
```python
>>> points_layer.text
['class: cat, condifidence: 0.95', 'class: dog, condifidence: 0.5']
```

**Comment by sofroniewn** [comment]
Closed by #1374
